test/io-tests/test_io.py

```diff
@@ -0,0 +1,220 @@
+    return result.stdout.decode("utf-8")
+
+
+@contextlib.contextmanager
+def run(configpath, stdin=None, moreenv=None):
+    "Run PostgREST."
+    env = os.environ
+    if moreenv:
+        env = {**env, **moreenv}
+
+    command = ["postgrest", configpath]
+    process = subprocess.Popen(command, stdin=subprocess.PIPE, env=env)
+
+    try:
+        if stdin:
+            process.stdin.write(stdin)
+        process.stdin.close()
+
+        waitfor200(BASEURL)
+        yield BASEURL
+    finally:
+        process.kill()
+        process.wait()
+
+
+def waitfor200(url):
+    for i in range(10):
+        try:
+            response = requests.get(url, timeout=0.1)
+
+            if response.status_code == 200:
+                return
+        except requests.ConnectionError:
+            pass
+
+        time.sleep(0.1)
+
+    raise TimeOutException()
+
+
+def test_expected_config(expectedconfig):
+    """
+    Configs as dumped by PostgREST should match an expected output.
+
+    Used to test default values, config aliases and environment variables. The
+    expected output for each file in 'configs', if available, is found in the
+    'configs/expected' directory.
+
+    """
+    expected = (basedir / "configs" / "expected" / expectedconfig).read_text()
+    assert dumpconfig(basedir / "configs" / expectedconfig) == expected
+
+
+def test_stable_config(configpath):
+    """
+    A dumped, re-read and re-dumped config should match the dumped config.
+
+    Note: only dump vs. re-dump must be equal, as the original config file might
+    be different because of default values, whitespace, and quoting.
+
+    """
+    env = {"ROLE_CLAIM_KEY": '."https://www.example.com/roles"[0].value'}
+    dumped = dumpconfig(configpath, moreenv=env)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        tmpconfigpath = pathlib.Path(tmpdir, "config")
+        tmpconfigpath.write_text(dumped)
+        redumped = dumpconfig(tmpconfigpath, moreenv=env)
+
+    assert dumped == redumped
+
+
+def test_read_secret_from_file(secretpath):
+    if secretpath.suffix == ".b64":
+        configfile = basedir / "configs" / "base64-secret-from-file.config"
+    else:
+        configfile = basedir / "configs" / "secret-from-file.config"
+
+    secret = secretpath.read_bytes()
+
+    jwt = secretpath.with_suffix(".jwt").read_text()
+    headers = {"Authorization": f"Bearer {jwt}"}
+
+    with run(configfile, stdin=secret) as url:
+        response = requests.get(f"{url}/authors_only", headers=headers)
+        assert response.status_code == 200
+
+
+def test_read_dburi_from_file_without_eol():
+    with run(dburifromfileconfig, stdin=dburi) as url:
+        response = requests.get(f"{url}/")
+        assert response.status_code == 200
+
+
+def test_read_dburi_from_file_with_eol():
+    with run(dburifromfileconfig, stdin=dburi + b"\n") as url:
+        response = requests.get(f"{url}/")
+        assert response.status_code == 200
+
+
+def test_role_claim_key(roleclaimcase):
```

Comment:
I'm trying to review the new tests as they come in and realized it's hard to look at one test with everything that goes into it. Currently the file is organized as:

```
var1 =
var2 = 
var3 =

fixture1
fixture2
fixture3

test1 
test2
test3
```

Now that the file is getting longer, I need to scroll more and more to get one test together. I think most the variables and fixtures belong to a small group of tests. What do you think about the following?

```
var1 =
fixture1
test1

var2 =
fixture2
test2

var3 =
fixture3
test3
```

A lot less scrolling needed to look at those tests!

Reply:
Yes, will refactor as everything comes together!